### ISSUE TITLE
Enumerate loaded runtimes from C++

### ISSUE THREAD

**Issue by quoctruong** [dotnet/coreclr#10568]
Hi,

I'm writing a C++ application that tries to get the `ICLRRuntimeInfo` from a .NET Core process through the following code:
```
HANDLE procHandle = OpenProcess(PROCESS_ALL_ACCESS, FALSE, <process ID>);
ICLRMetaHost *pMetaHost = NULL;
CLRCreateInstance(CLSID_CLRMetaHost, IID_ICLRMetaHost, (LPVOID*)&pMetaHost);
IEnumUnknown *pEnumUnknown = NULL;
pMetaHost->EnumerateLoadedRuntimes(procHandle, &pEnumUnknown);
```

However, the pEnumUnknown returned does not contain any value (this code works if the process is full .NET). Is this not supported?

Ultimately, I'm trying to get the `ICLRRuntimeInfo` so I can get a `ICorDebug` object by calling `GetInterface` on the `ICLRRuntimeInfo` object.

**Comment by gkhanna79** [comment]
These interfaces are not supported for .NET Core.

**Comment by quoctruong** [comment]
@gkhanna79, thanks for the response. Is there any way I can get the `ICorDebug` object for CoreCLR? Or is there any equivalent debugging interface?

**Comment by gkhanna79** [comment]
I don't believe so. @noahfalk do you know know?



Thanks,
Gaurav


-------- Original message --------

**Comment by noahfalk** [comment]
@mikem8361 

Mike added some startup APIs for coreclr debugging a while back:
https://github.com/dotnet/coreclr/pull/2507/commits/178d90989a6b95981c5a094bb418fa4d59918634

I that API surface is still current and the ongoing work has been to fix some bugs. Mike would know the most up-to-date info.

**Comment by quoctruong** [comment]
It seems I can get access to most of the debugging API by loading some functions from dbgshim.dll. @mikem8361, is this the recommended way for writing a debugger?

**Comment by mikem8361** [comment]
Dbgshim now provides some new launch/attach APIs that work across all platforms:

CreateProcessForLaunch
ResumeProcess
CloseResumeHandle
RegisterForRuntimeStartup
UnregisterForRuntimeStartup

For launch:

1) CreateProcessForLaunch(commandLine, TRUE, environment, currentDirectory, &processId, &resumeHandle);
2) RegisterForRuntimeStartup(processId, CallbackFunction, parameter, &unregisterToken);
3) ResumeProcess(resumeHandle);
4) CloseResumeHandle(resumeHandle);

Inside the CallbackFunction:

void
CallbackFunction(IUnknown* punk, void* parameter, HRESULT hr)
{
      if (hr == S_OK) {
         ICorDebug* pCordb = (ICorDebug *)punk;
        //hr = punk->QueryInterface(IID_ICorDebug, (void **)&pCordb);

        hr = pCordb->Initialize();

        ManagedCallback *callback = new ManagedCallback();
        hr = pCordb->SetManagedHandler(callback);

        ICorDebugProcess *process;
        hr = pCordb->DebugActiveProcess(processId, FALSE, &process);
      }
      UnregisterForRuntimeStartup(unregisterToken);
}

For attach: skip the CreateProcessForLaunch/ResumeProcess, etc. steps and just use RegisterForRuntimeStartup.

**Comment by quoctruong** [comment]
@mikem8361, thank you. This is super helpful!

**Comment by danmosemsft** [comment]
@quoctruong is this resolved, did you have success?

**Comment by quoctruong** [comment]
Hi @mikem8361, on my Ubuntu 14.04, I was able to attach to a process and create an ICorDebug object using your callback function. However, when I initialize the object, I ran into some segmentation fault problem. In particular, I got the following error (from gdb):
```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff778aa53 in EnvironGetenv (name=0x7fffec00cb50 "COMPlus_StressLog", copyValue=0)
   at /coreclr/src/pal/src/misc/environ.cpp:928
928	   for (int i = 0; palEnvironment[i] != nullptr; ++i)
```
Not sure if this is correct but it seems that the [InitializeCommon function](https://github.com/dotnet/coreclr/blob/master/src/debug/di/rsmain.cpp#L481) is trying to get the environment value StressLog and that somehow results in this problem?

Sorry to reopen the question again. Thanks!
